src/inference.rs

```diff
@@ -162,200 +172,230 @@ impl<J: Jet> TryFrom<&CommitNode<J>> for NodeType {
+            // Multiple children -- now we have nontrivial unification and may fail.
+            CommitNodeInner::Comp(ref lchild, ref rchild) => {
+                let arrow = UnificationArrow {
+                    source: lchild.arrow.source.clone(),
+                    target: rchild.arrow.target.clone(),
+                };
+                unify(
+                    lchild.arrow.target.clone(),
+                    rchild.arrow.source.clone(),
+                    "Composition: Left target = right source",
+                )?;
+                Ok(arrow)
+            }
+            CommitNodeInner::Case(ref lchild, ref rchild)
+            | CommitNodeInner::AssertL(ref lchild, ref rchild)
+            | CommitNodeInner::AssertR(ref lchild, ref rchild) => {
+                let a = naming.free_variable();
+                let b = naming.free_variable();
+                let c = naming.free_variable();
+
+                let sum_a_b = VariableType::Sum(a.clone(), b.clone());
+                let prod_sum_a_b_c = VariableType::Product(Variable::bound(sum_a_b), c.clone());
+                let source = Variable::bound(prod_sum_a_b_c);
+
+                let target = match node {
+                    CommitNodeInner::AssertL(..) => {
                         bind(
-                            &find_root(left_arrow.source.clone()),
-                            VariableType::Product(a, c.clone()),
+                            &find_root(lchild.arrow.source.clone()),
+                            VariableType::Product(a, c),
                             "Case: Left source = A × C",
                         )?;
-                        unify(
-                            arrow.target.clone(),
-                            left_arrow.target.clone(),
-                            "Cannot fail",
+                        lchild.arrow.target.clone()
+                    }
+                    CommitNodeInner::AssertR(..) => {
+                        bind(
+                            &find_root(rchild.arrow.source.clone()),
+                            VariableType::Product(b, c),
+                            "Case: Right source = B × C",
                         )?;
+                        rchild.arrow.target.clone()
                     }
-                    if let CommitNodeInner::AssertL(_, _) = node {
-                    } else {
+                    CommitNodeInner::Case(..) => {
                         bind(
-                            &find_root(right_arrow.source.clone()),
+                            &find_root(lchild.arrow.source.clone()),
+                            VariableType::Product(a, c.clone()),
+                            "Case: Left source = A × C",
+                        )?;
+                        bind(
+                            &find_root(rchild.arrow.source.clone()),
                             VariableType::Product(b, c),
                             "Case: Right source = B × C",
                         )?;
```

Comment:
We are calling `find_root` twice here because of the changes in `bind`. Originally `bind` didn't find the root because some arguments were guaranteed to be roots. This is no longer the case in this PR because these cases are handled differently.

Reply:
Ah, good catch! Yep, I will remove all the `find_root` calls here.